papers/bradley_dice/55_bradley_dice.rst

```diff
@@ -0,0 +1,584 @@
+Users can quickly save notes into the job document and then search those notes, which is useful for high throughput studies that require some manual investigation (e.g. reviewing plots).
+
+Performance Enhancements
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+In early 2021, a significant portion of the codebase was profiled and refactored to improve performance and these improvements were released in **signac** 1.6.0 and **signac-flow** 0.12.0.
+As a result of these changes, large **signac** projects saw 4-7x speedups for operations such as iterating over the jobs in a project compared to the 1.5.0 release of **signac**.
+Similarly, performance of a sample workflow that checks status, runs, and submits a FlowProject with 1,000 jobs, 3 operations, and 2 label functions improved roughly 4x compared to **signac-flow** 0.11.0.
+These improvements allow **signac** to scale to ~100,000 jobs.
+
+In **signac**, the core of the ``Project`` and ``Job`` classes were refactored to support lazy attribute access and delayed initialization, which greatly reduces the total amount of disk I/O by waiting until data is actually requested by the user.
+Other improvements include early exits in functions, reducing the number of required system calls with smarter usage of the ``os`` library, and switching to algorithms that operate in constant time (:math:`O(1)`) instead of linear time (:math:`O(N_{jobs})`).
+Optimizations were identified by profiling the performance of common operations on small and large real-world projects with cProfile and visualized with snakeviz :cite:`snakeviz`.
+
+..
+    TODO: include a graph of performance from 1.0 to now
+
+Similarly, performance enhancements were also made in the **signac-flow** package.
+Some of the optimizations identified include lazy evaluation of run commands and directives, and caching of job status information.
+In addition, the improvements in **signac** such as faster iteration over large **signac** projects used in **signac-flow** made **signac-flow**'s primary functions — checking project status, executing operations, and submitting operations to a cluster — significantly faster.
+
+Improved User Output
+~~~~~~~~~~~~~~~~~~~~
+
+**Workflow graph detection:**
+The preconditions and postconditions of operations in a **signac-flow** ``FlowProject`` implicitly define a graph. For example, if the operation "analyze" depends on the operation "simulate" via the precondition ``@FlowProject.pre.after(simulate)``, then there is a directed edge from "simulate" to "analyze."
+This graph can now be detected from the workflow conditions and returned in a NetworkX compatible format for display or inspection.
+
+**Templated status output:**
+Querying the status of a **signac-flow** project now has many options controlling the information displayed and has been templated to allow for raw, Markdown, or HTML output. In doing so, the output has also become cleaner and compatible with external tools.
```

Comment:
What does "raw" mean in this context? JSON-formatted?

Reply:
Good catch. This means "plain text" for the terminal display, I believe. I'll adjust:
```suggestion
Querying the status of a **signac-flow** project now has many options controlling the information displayed and has been templated to allow for plain text, Markdown, or HTML output. In doing so, the output has also become cleaner and compatible with external tools.
```